File: tests/test_clip.py
# SPDX-License-Identifier: Apache-2.0
# Copyright Contributors to the OpenTimelineIO project

import unittest
import pytest

import opentimelineio as otio
import opentimelineio.test_utils as otio_test_utils


class ClipTests(unittest.TestCase, otio_test_utils.OTIOAssertions):

    def test_cons(self):
        name = "test"
        rt = otio.opentime.RationalTime(5, 24)
        tr = otio.opentime.TimeRange(rt, rt)
        mr = otio.schema.ExternalReference(
            available_range=otio.opentime.TimeRange(
                rt,
                otio.opentime.RationalTime(10, 24)
            ),
            target_url="/var/tmp/test.mov"
        )

        cl = otio.schema.Clip(
            name=name,
            media_reference=mr,
            source_range=tr,
            # transition_in
            # transition_out
        )
        assert cl.name == name
        assert cl.source_range == tr
        self.assertIsOTIOEquivalentTo(cl.media_reference, mr)

        encoded = otio.adapters.otio_json.write_to_string(cl)
        decoded = otio.adapters.otio_json.read_from_string(encoded)
        self.assertIsOTIOEquivalentTo(cl, decoded)

    def test_each_clip(self):
        cl = otio.schema.Clip(name="test_clip")
        assert list(cl.each_clip()) == [cl]

    def test_str(self):
        cl = otio.schema.Clip(name="test_clip")

        self.assertMultiLineEqual(
            str(cl),
            'Clip('
            '"test_clip", '
            'MissingReference(\'\', None, None, {}), '
            'None, {})'
        )
        self.assertMultiLineEqual(
            repr(cl),
            'otio.schema.Clip('
            "name='test_clip', "
            'media_reference={}, '
            'source_range=None, '
            'metadata={{}}'
            ')'.format(
                repr(cl.media_reference)
            )
        )

    def test_str_with_filepath(self):
        cl = otio.schema.Clip(
            name="test_clip",
            media_reference=otio.schema.ExternalReference(
                "/var/tmp/foo.mov"
            )
        )
        self.assertMultiLineEqual(
            str(cl),
            'Clip('
            '"test_clip", ExternalReference("/var/tmp/foo.mov"), None, {}'
            ')'
        )
        self.assertMultiLineEqual(
            repr(cl),
            'otio.schema.Clip('
            "name='test_clip', "
            "media_reference=otio.schema.ExternalReference("
            "target_url='/var/tmp/foo.mov'"
            "), "
            'source_range=None, '
            'metadata={}'
            ')'
        )

    def test_ranges(self):
        tr = otio.opentime.TimeRange(
            # 1 hour in at 24 fps
            start_time=otio.opentime.RationalTime(86400, 24),
            duration=otio.opentime.RationalTime(200, 24)
        )

        cl = otio.schema.Clip(
            name="test_clip",
            media_reference=otio.schema.ExternalReference(
                "/var/tmp/foo.mov",
                available_range=tr
            )
        )

        assert cl.duration() == cl.trimmed_range().duration
        assert cl.duration() == tr.duration
        assert cl.trimmed_range() == tr
        assert cl.available_range() == tr
        assert cl.trimmed_range() is not tr
        assert cl.available_range() is not tr

        cl.source_range = otio.opentime.TimeRange(
            # 1 hour + 100 frames
            start_time=otio.opentime.RationalTime(86500, 24),
            duration=otio.opentime.RationalTime(50, 24)
        )

        assert cl.duration() != tr.duration
        assert cl.trimmed_range() != tr
        assert cl.duration() == cl.source_range.duration
        assert cl.duration() is not cl.source_range.duration
        assert cl.trimmed_range() == cl.source_range
        assert cl.trimmed_range() is not cl.source_range

    def test_available_image_bounds(self):
        available_image_bounds = otio.schema.Box2d(
            otio.schema.V2d(0.0, 0.0),
            otio.schema.V2d(16.0, 9.0)
        )

        media_reference = otio.schema.ExternalReference(
            "/var/tmp/foo.mov",
            available_image_bounds=available_image_bounds
        )

        cl = otio.schema.Clip(
            name="test_available_image_bounds",
            media_reference=media_reference
        )

        assert available_image_bounds == cl.available_image_bounds
        assert cl.available_image_bounds == media_reference.available_image_bounds

        assert 0.0 == cl.available_image_bounds.min.x
        assert 0.0 == cl.available_image_bounds.min.y
        assert 16.0 == cl.available_image_bounds.max.x
        assert 9.0 == cl.available_image_bounds.max.y

        # test range exceptions
        cl.media_reference.available_image_bounds = None
        with self.assertRaises(otio.exceptions.CannotComputeAvailableRangeError):
            cl.available_range()

    def test_ref_default(self):
        cl = otio.schema.Clip()
        self.assertIsOTIOEquivalentTo(
            cl.media_reference,
            otio.schema.MissingReference()
        )

        cl.media_reference = None
        self.assertIsOTIOEquivalentTo(
            cl.media_reference,
            otio.schema.MissingReference()
        )

        cl.media_reference = otio.schema.ExternalReference()
        self.assertIsOTIOEquivalentTo(
            cl.media_reference,
            otio.schema.ExternalReference()
        )

    def test_multi_ref(self):
        cl = otio.schema.Clip()

        assert otio.schema.Clip.DEFAULT_MEDIA_KEY == cl.active_media_reference_key
        self.assertIsOTIOEquivalentTo(
            cl.media_reference,
            otio.schema.MissingReference()
        )

        mrs = cl.media_references()
        self.assertIsOTIOEquivalentTo(
            mrs[otio.schema.Clip.DEFAULT_MEDIA_KEY],
            otio.schema.MissingReference()
        )

        cl.set_media_references(
            {
                otio.schema.Clip.DEFAULT_MEDIA_KEY: otio.schema.ExternalReference(),
                "high_quality": otio.schema.GeneratorReference(),
                "proxy_quality": otio.schema.ImageSequenceReference(),
            },
            otio.schema.Clip.DEFAULT_MEDIA_KEY
        )
        mrs = cl.media_references()
        self.assertIsOTIOEquivalentTo(
            mrs[otio.schema.Clip.DEFAULT_MEDIA_KEY],
            otio.schema.ExternalReference()
        )
        self.assertIsOTIOEquivalentTo(
            mrs["high_quality"],
            otio.schema.GeneratorReference()
        )
        self.assertIsOTIOEquivalentTo(
            mrs["proxy_quality"],
            otio.schema.ImageSequenceReference()
        )

        cl.active_media_reference_key = "high_quality"
        self.assertIsOTIOEquivalentTo(
            cl.media_reference,
            otio.schema.GeneratorReference()
        )
        assert cl.active_media_reference_key == "high_quality"

        # we should get an exception if we try to use a key that is
        # not in the media_references
        with pytest.raises(ValueError):
            cl.active_media_reference_key = "cloud"

        assert cl.active_media_reference_key == "high_quality"

        # we should also get an exception if we set the references without
        # the active key
        with pytest.raises(ValueError):
            cl.set_media_references(
                {
                    "cloud": otio.schema.ExternalReference()
                },
                "high_quality"
            )

        assert cl.active_media_reference_key == "high_quality"

        # we should also get an exception if we set the references with
        # an empty key
        with pytest.raises(ValueError):
            cl.set_media_references(
                {
                    "": otio.schema.ExternalReference()
                },
                ""
            )

        assert cl.active_media_reference_key == "high_quality"

        # setting the references and the active key should resolve the problem
        cl.set_media_references(
            {
                "cloud": otio.schema.ExternalReference()
            },
            "cloud"
        )

        assert cl.active_media_reference_key == "cloud"
        self.assertIsOTIOEquivalentTo(
            cl.media_reference, otio.schema.ExternalReference())


if __name__ == '__main__':
    unittest.main()

File: tests/test_clip.cpp
// SPDX-License-Identifier: Apache-2.0
// Copyright Contributors to the OpenTimelineIO project

#include "utils.h"

#include <opentimelineio/clip.h>
#include <opentimelineio/deserialization.h>
#include <opentimelineio/externalReference.h>
#include <opentimelineio/freezeFrame.h>
#include <opentimelineio/linearTimeWarp.h>
#include <opentimelineio/marker.h>
#include <opentimelineio/missingReference.h>
#include <opentimelineio/serializableCollection.h>
#include <opentimelineio/timeline.h>

#include <iostream>

using namespace OTIO_NS;

int
main(int argc, char** argv)
{
    Tests tests;

    tests.add_test("test_cons", [] {
        std::string  name = "test";
        RationalTime rt(5, 24);
        TimeRange    tr(rt, rt);

        SerializableObject::Retainer<ExternalReference> mr(
            new ExternalReference);
        mr->set_available_range(TimeRange(rt, RationalTime(10, 24)));
        mr->set_target_url("/var/tmp/test.mov");

        SerializableObject::Retainer<Clip> cl(new Clip);
        cl->set_name(name);
        cl->set_media_reference(mr);
        cl->set_source_range(tr);
        assertEqual(cl->name(), name);
        assertEqual(cl->source_range().value(), tr);

        std::string encoded = cl->to_json_string();
        SerializableObject::Retainer<SerializableObject> decoded(
            SerializableObject::from_json_string(encoded));
        assertTrue(cl->is_equivalent_to(*decoded));
    });

    tests.add_test("test_ranges", [] {
        TimeRange tr(
            // 1 hour in at 24 fps
            RationalTime(86400, 24),
            RationalTime(200, 24));

        SerializableObject::Retainer<Clip> cl(new Clip("test_clip"));
        SerializableObject::Retainer<ExternalReference> mr(
            new ExternalReference);
        mr->set_target_url("/var/tmp/test.mov");
        mr->set_available_range(tr);
        cl->set_media_reference(mr);
        assertEqual(cl->duration(), cl->trimmed_range().duration());
        assertEqual(cl->duration(), tr.duration());
        assertEqual(cl->trimmed_range(), tr);
        assertEqual(cl->available_range(), tr);

        cl->set_source_range(TimeRange(
            // 1 hour + 100 frames
            RationalTime(86500, 24),
            RationalTime(50, 24)));
        assertNotEqual(cl->duration(), tr.duration());
        assertNotEqual(cl->trimmed_range(), tr);
        assertEqual(cl->duration(), cl->source_range()->duration());
        assertEqual(cl->trimmed_range(), cl->source_range().value());
    });

    tests.add_test("test_clip_v1_to_v2_null", [] {
        OTIO_NS::ErrorStatus           status;
        SerializableObject::Retainer<> so =
            SerializableObject::from_json_string(
                R"(
            {
                "OTIO_SCHEMA": "Clip.1",
                "media_reference": null
            })",
                &status);

        assertFalse(is_error(status));

        const Clip* clip = dynamic_cast<const Clip*>(so.value);
        assertNotNull(clip);

        const MissingReference* media_ref =
            dynamic_cast<MissingReference*>(clip->media_reference());
        assertNotNull(media_ref);
    });

    tests.add_test("test_clip_v1_to_v2", [] {
        OTIO_NS::ErrorStatus           status;
        SerializableObject::Retainer<> so =
            SerializableObject::from_json_string(
                R"(
            {
                "OTIO_SCHEMA": "Clip.1",
                "media_reference": {
                    "OTIO_SCHEMA": "ExternalReference.1",
                    "target_url": "unit_test_url",
                    "available_range": {
                        "OTIO_SCHEMA": "TimeRange.1",
                        "duration": {
                            "OTIO_SCHEMA": "RationalTime.1",
                            "rate": 24,
                            "value": 8
                        },
                        "start_time": {
                            "OTIO_SCHEMA": "RationalTime.1",
                            "rate": 24,
                            "value": 10
                        }
                    }
                }
            })",
                &status);

        assertFalse(is_error(status));

        const Clip* clip = dynamic_cast<const Clip*>(so.value);
        assertNotNull(clip);

        const ExternalReference* media_ref =
            dynamic_cast<ExternalReference*>(clip->media_reference());
        assertNotNull(media_ref);
        assertEqual(
            clip->active_media_reference_key().c_str(),
            Clip::default_media_key);

        assertEqual(media_ref->target_url().c_str(), "unit_test_url");
        assertEqual(media_ref->available_range()->duration().value(), 8);
        assertEqual(media_ref->available_range()->duration().rate(), 24);
        assertEqual(media_ref->available_range()->start_time().value(), 10);
        assertEqual(media_ref->available_range()->start_time().rate(), 24);

        auto mediaReferences = clip->media_references();
        auto found           = mediaReferences.find(Clip::default_media_key);
        assertFalse(found == mediaReferences.end());
    });

    tests.add_test("test_clip_media_representation", [] {
        static constexpr auto time_scalar = 1.5;

        SerializableObject::Retainer<LinearTimeWarp> ltw(new LinearTimeWarp(
            LinearTimeWarp::Schema::name,
            LinearTimeWarp::Schema::name,
            time_scalar));
        std::vector<Effect*>                         effects = { ltw };

        static constexpr auto red = Marker::Color::red;

        SerializableObject::Retainer<Marker> m(
            new Marker(LinearTimeWarp::Schema::name, TimeRange(), red));
        std::vector<Marker*> markers = { m };

        static constexpr auto high_quality  = "high_quality";
        static constexpr auto proxy_quality = "proxy_quality";

        SerializableObject::Retainer<MediaReference> media(
            new ExternalReference());

        SerializableObject::Retainer<Clip> clip(new Clip(
            "unit_clip",
            media,
            std::nullopt,
            AnyDictionary(),
            effects,
            markers,
            high_quality));

        assertEqual(clip->active_media_reference_key().c_str(), high_quality);
        assertEqual(clip->media_reference(), media.value);
        assertEqual(clip->active_media_reference_key().c_str(), high_quality);

        SerializableObject::Retainer<MediaReference> ref1(
            new ExternalReference());
        SerializableObject::Retainer<MediaReference> ref2(
            new ExternalReference());
        SerializableObject::Retainer<MediaReference> ref3(
            new ExternalReference());

        clip->set_media_references(
            { { Clip::default_media_key, ref1 },
              { high_quality, ref2 },
              { proxy_quality, ref3 } },
            high_quality);

        assertEqual(clip->media_reference(), ref2.value);

        clip->set_active_media_reference_key(proxy_quality);
        assertEqual(clip->media_reference(), ref3.value);

        clip->set_active_media_reference_key(Clip::default_media_key);
        assertEqual(clip->media_reference(), ref1.value);

        // setting the active reference to a key that does not exist
        // should return an error
        OTIO_NS::ErrorStatus error;
        clip->set_active_media_reference_key("cloud", &error);
        assertTrue(is_error(error));
        assertEqual(
            error.outcome,
            OTIO_NS::ErrorStatus::MEDIA_REFERENCES_DO_NOT_CONTAIN_ACTIVE_KEY);
        assertEqual(clip->media_reference(), ref1.value);

        // setting the references that doesn't have the active key should
        // also generate an error
        SerializableObject::Retainer<MediaReference> ref4(
            new ExternalReference());

        OTIO_NS::ErrorStatus error2;
        clip->set_media_references(
            { { "cloud", ref4 } },
            high_quality,
            &error2);
        assertTrue(is_error(error2));
        assertEqual(
            error2.outcome,
            OTIO_NS::ErrorStatus::MEDIA_REFERENCES_DO_NOT_CONTAIN_ACTIVE_KEY);

        assertEqual(clip->media_reference(), ref1.value);

        // setting an empty should also generate an error
        OTIO_NS::ErrorStatus error3;
        clip->set_media_references({ { "", ref4 } }, "", &error3);
        assertTrue(is_error(error3));
        assertEqual(
            error3.outcome,
            OTIO_NS::ErrorStatus::MEDIA_REFERENCES_CONTAIN_EMPTY_KEY);

        // setting the references and the active key at the same time
        // should work
        clip->set_media_references({ { "cloud", ref4 } }, "cloud");
        assertEqual(clip->media_reference(), ref4.value);

        // basic test for an effect
        assertEqual(clip->effects().size(), effects.size());
        auto effect = dynamic_cast<OTIO_NS::LinearTimeWarp*>(
            clip->effects().front().value);
        assertEqual(effect->time_scalar(), time_scalar);

        // basic test for a marker
        assertEqual(clip->markers().size(), markers.size());
        auto marker =
            dynamic_cast<OTIO_NS::Marker*>(clip->markers().front().value);
        assertEqual(marker->color().c_str(), red);
    });

    // test to ensure null error_status pointers are correctly handled
    tests.add_test("test_error_ptr_null", [] {
        // tests for no image bounds on media reference on clip
        SerializableObject::Retainer<Clip> clip(new Clip);

        // check that there is an error, and that it's the correct error
        OTIO_NS::ErrorStatus mr_bounds_error;
        clip->available_image_bounds(&mr_bounds_error);
        assertTrue(is_error(mr_bounds_error));
        assertEqual(
            mr_bounds_error.outcome,
            OTIO_NS::ErrorStatus::CANNOT_COMPUTE_BOUNDS);

        // check that if null ptr, nothing happens
        OTIO_NS::ErrorStatus* null_test = nullptr;

        assertEqual(
            clip->available_image_bounds(null_test),
            std::optional<IMATH_NAMESPACE::Box2d>());
    });

    // test to ensure null error_status pointers are correctly handled
    // when there's no media reference
    tests.add_test("test_error_ptr_null_no_media", [] {
        SerializableObject::Retainer<Clip> clip(new Clip);

        // set media reference to empty
        Clip::MediaReferences empty_mrs;
        empty_mrs["empty"] = nullptr;
        clip->set_media_references(empty_mrs, "empty");

        OTIO_NS::ErrorStatus bounds_error_no_mr;
        clip->available_image_bounds(&bounds_error_no_mr);
        assertTrue(is_error(bounds_error_no_mr));

        assertEqual(
            bounds_error_no_mr.outcome,
            OTIO_NS::ErrorStatus::CANNOT_COMPUTE_BOUNDS);

        // std::cout<< "bounds error details: " << bounds_error_no_mr.details << std::endl;

        OTIO_NS::ErrorStatus* null_test_no_mr = nullptr;

        assertEqual(
            clip->available_image_bounds(null_test_no_mr),
            std::optional<IMATH_NAMESPACE::Box2d>());
    });

    tests.run(argc, argv);
    return 0;
}
